pandas/core/dtypes/cast.py

```diff
@@ -320,6 +320,12 @@ def maybe_cast_result(
     """
     dtype = obj.dtype
     dtype = maybe_cast_result_dtype(dtype, how)
+    # result_dtype = maybe_cast_result_dtype(dtype, how)
+    # if result_dtype is not None:
+    #     # we know what the result dtypes needs to be -> be more permissive in casting
+    #     # (eg ints with nans became floats)
+    #     cls = result_dtype.construct_array_type()
+    #     return cls._from_sequence(obj, dtype=result_dtype)
```

Comment:
This is related to the first case mentioned (when we *know* the resulting dtype, we should maybe force the result back, instead of relying on strict casting). 
So the above code is one possibility: we can change `maybe_cast_result_dtype` to only return a dtype if it knows what dtype should be and otherwise return None (instead of passing through the original type). This way, we can take a different path for "known" dtypes, vs when guessing the dtype.